smart-contract-extensions/contracts/unlock-examples/hooks/CodeRequiredHook.sol

```diff
@@ -0,0 +1,66 @@
+pragma solidity 0.5.16;
+
+import '@openzeppelin/contracts/cryptography/ECDSA.sol';
+import '@openzeppelin/contracts/access/roles/WhitelistAdminRole.sol';
+import '@openzeppelin/contracts/introspection/IERC1820Registry.sol';
+import 'hardlydifficult-ethereum-contracts/contracts/proxies/CallContract.sol';
+
+/**
+ * @notice Used with a Lock in order to require the user knows
+ * a code in order to buy.
+ * @dev Set this contract as the beneficiary of the lock.
```

Comment:
So here we set the CodeRequiredHook contract as the beneficiary, but I'm not seeing how changing the beneficiary(as per comments for proxyCall) would remove the code requirement? I would have thought we'd need to update the call to this hook from the purchase function.

Reply:
We check for a hook today by asking erc-1820 if the beneficiary has registered one.

Reply:
We could either make this contract the beneficiary, or instead have the beneficiary call erc1820 to say this is the hook to use.

The downside with the latter is the beneficiary's other locks will call the hook as well.  So the hook needs to check if it's the lock of interest... which is what motivated our desire to remove 1820 completely.

Reply:
Ah, right. So while the new hook contract is the beneficiary, there is a registered hook (itself). Changing the beneficary means the new beneficiary won't have (by default) a registered hook, and the purchase will proceed as a normal purchase.

One question about this: If all purchase calls have to check the erc-1820 contract for registered hooks, what sort of gas-penalty do users pay for this in the cases where a hook is not used?

Reply:
That's right.

It's a non-zero cost to use 1820 for sure, another motivation to change.